ena-submission/src/ena_deposition/check_external_visibility.py

```diff
@@ -0,0 +1,399 @@
+Add timestamp to project/sample/assembly table when first publicly visible
+1. Find all accessions that don't yet have a publicly visible timestamp
+2. Query ENA and NCBI for these accessions
+3. If publicly visible, update the timestamp in the database
+4. Repeat periodically
+"""
+
+import logging
+import threading
+import time
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from http import HTTPStatus
+from typing import Any
+
+import pytz
+import requests
+from psycopg2.pool import SimpleConnectionPool
+
+from ena_deposition.config import Config
+from ena_deposition.submission_db_helper import (
+    AssemblyTableEntry,
+    ProjectTableEntry,
+    SampleTableEntry,
+    Status,
+    TableName,
+    db_init,
+    find_conditions_in_db,
+    update_db_where_conditions,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class EntityType(Enum):
+    PROJECT = "project"
+    SAMPLE = "sample"
+    ASSEMBLY = "assembly"
+
+
+@dataclass
+class ColumnCheckConfig:
+    """Configuration for checking a specific table column"""
+
+    table_name: TableName
+    entry_class: type
+    id_fields: list[str]
+    visibility_column: str
+    accession_field: str  # Field prefix in the result dict (e.g., "insdc_accession_full")
+    checker_class: type  # Which visibility checker to use
+    check_all_segments: bool = False  # Whether to check all keys starting with accession_field
+
+
+class VisibilityChecker(ABC):
+    """Abstract base class for visibility checkers"""
+
+    @abstractmethod
+    def check_visibility(self, config: Config, accession: str) -> datetime | None:
+        """Check if an accession is publicly visible"""
+
+
+class ENAVisibilityChecker(VisibilityChecker):
+    """Checker for ENA visibility"""
+
+    def check_visibility(self, config: Config, accession: str) -> datetime | None:
+        file_type = "xml" if accession.startswith(("PRJ", "SAM", "GCA")) else "embl"
+        response = requests.get(
+            f"https://www.ebi.ac.uk/ena/browser/api/{file_type}/{accession}",
+            allow_redirects=False,
+            timeout=config.ena_http_timeout_seconds,
+        )
+        if response.status_code == HTTPStatus.OK:
+            return datetime.now(pytz.UTC)
+        return None
+
+
+GcaCacheKey = tuple[str] | tuple[str, str] | tuple[str, str, str]
+
+gca_cache: dict[GcaCacheKey, bool] = {}
+
+
+def _check_and_cache_gca(config: Config, path_segments: GcaCacheKey) -> bool:
+    """
+    Helper function to check ENA for a GCA accession part and cache the result.
+    Returns True if found and caches True, False if not found and caches False.
+    """
+    cache_key = path_segments
+
+    if cache_key in gca_cache:
+        return gca_cache[cache_key]
+
+    url_path = "/".join(path_segments)
+    response = requests.get(
+        f"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/{url_path}/",
```

Comment:
just out of curiosity why did you choose to add caching? it makes the code a bit more complex so I assume NCBI was rate-limiting or sth - maybe you could add a comment

Reply:
Im fine to keep this :-)